src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,232 @@
+by ext; simp [circulant]
+
+lemma fin.transpose_circulant : ∀ {n} (v : fin n → α), (circulant v)ᵀ =  circulant (λ i, v (-i))
+| 0     := dec_trivial
+| (n+1) := transpose_circulant
+
+lemma fin.conj_transpose_circulant [has_star α] :
+  ∀ {n} (v : fin n → α), (circulant v)ᴴ = circulant (star (λ i, v (-i)))
+| 0     := dec_trivial
+| (n+1) := conj_transpose_circulant
+
+lemma map_circulant [has_sub I] (v : I → α) (f : α → β) :
+  (circulant v).map f = circulant (λ i, f (v i)) :=
+by ext; simp [circulant]
+
+lemma circulant_neg [has_neg α] [has_sub I] (v : I → α) :
+  circulant (- v) = - circulant v :=
+by ext; simp [circulant]
+
+lemma circulant_add [has_add α] [has_sub I] (v w : I → α) :
+  circulant (v + w) = circulant v + circulant w :=
+by ext; simp [circulant]
+
+lemma circulant_sub [has_sub α] [has_sub I] (v w : I → α) :
+  circulant (v - w) = circulant v - circulant w :=
+by ext; simp [circulant]
+
+lemma circulant_mul [semiring α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply],
+  congr' 2,
+  abel
+end
+
+lemma fin.circulant_mul [semiring α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w)
+| 0     := dec_trivial
+| (n+1) := circulant_mul
+
+/-- circulant matrices commute in multiplication under certain condations. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] {k : R} {v : I → α} :
+  circulant (k • v) = k • circulant v :=
+by {ext, simp [circulant]}
+
+@[simp] lemma circulant_zero [has_zero α] [has_sub I]:
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp [circulant]
+
+@[simp] lemma circulant_single [has_zero α] [has_one α] [decidable_eq I] [add_group I]:
+  circulant (pi.single 0 1 : I → α)= (1 : matrix I I α) :=
+by { ext i j, simp only [circulant, one_apply, pi.single_apply, sub_eq_zero], congr }
+
+/-- The identity matrix is a circulant matrix. -/
+lemma one_eq_circulant [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
```

Comment:
I guess you can generalise this to scalar matrices. (This version is also useful.)

Reply:
I am not sure how to prove this though:
```
/-- Any scalar matrix is a circulant matrix. -/
lemma scalar_eq_circulant [has_zero α] [has_one α] [has_add α] [decidable_eq I] [add_group I] 
  (k : ℕ):
  (k : matrix I I α) = circulant (λ i, ite (i = 0) k 0) :=
by sorry
```
@jcommelin

Reply:
No, I meant using `matrix.scalar`.